libbeat/processors/uuid/generator/elasticsearch/generator.go

```diff
@@ -0,0 +1,111 @@
+// Licensed to Elasticsearch B.V. under one or more contributor
+// license agreements. See the NOTICE file distributed with
+// this work for additional information regarding copyright
+// ownership. Elasticsearch B.V. licenses this file to you under
+// the Apache License, Version 2.0 (the "License"); you may
+// not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+package elasticsearch
+
+import (
+	"encoding/base64"
+	"math/rand"
+	"sync"
+	"time"
+)
+
+var (
+	sequenceNumber uint32
+	lastTimestamp  *time.Time
+	mac            []byte
+	mu             sync.Mutex
+)
+
+func init() {
+	m, err := getSecureMungedMACAddress()
+	if err != nil {
+		panic(err)
+	}
+	mac = m
+	sequenceNumber = rand.Uint32()
+}
+
+// GetBase64UUID returns a base64-encoded, randomly-generated, but roughly ordered (over time), unique
+// ID. The algorithm used to generate the ID is the same as used by Elasticsearch.
+// See https://github.com/elastic/elasticsearch/blob/a666fb2266/server/src/main/java/org/elasticsearch/common/TimeBasedUUIDGenerator.java
+func GetBase64UUID() string {
```

Comment:
We should allow for some level of polymorphism. Which means interfaces and concrete types (e.g. consider singleton pattern if normal function would do the trick). How about:

```
type IDGenerator interface {
  NextID() string
}
```

with 

```
type esTimeBasedUUIDGenerator struct {}

// singleton pattern trick if type is only required to fulfill an interface declaration
var _esTimeBasedUUIDGenerator IDGenerator = (*esTimeBasedUUIDGenerator)(nil)

func TimeBasedUUIDGenerator() IDGenerator {
  return _esTimeBasedUUIDGenerator
}

func (_ *esTimeBasedUUIDGenerator) NextID() string {
  ...
}
```

As we don't have too many generators, I think all these types and definition can easily live in the generator package.

Now the processor can keep track of supported generators via:

```
var generators = map[string]func() IDGenerator {
 "elasticsearch_uuid": TimeBasedUUIDGenerator,
 ...
}
```